marketplace/deployer_util/validate.py

```diff
@@ -0,0 +1,185 @@
+#!/usr/bin/env python2
+#
+# Copyright 2018 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+import json
+import yaml
+
+from argparse import ArgumentParser
+from bash_util import Command
+from dict_util import deep_get
+from yaml_util import load_resources_yaml
+from yaml_util import load_yaml
+
+DEPLOY_INFO_ANNOTATION = "marketplace.cloud.google.com/deploy-info"
+GOOGLE_MARKETPLACE = "x-google-marketplace"
+
+_PROG_HELP="""
+Validates and test the deployer image.
+"""
+
+class ValidationException(Exception):
+  pass
+
+def main():
+  parser = ArgumentParser(description=_PROG_HELP)
+  parser.add_argument('--metadata', help='Path to metadata file for validation. If it is not present, the validation on the metadata file will be skipped.',)
+  args = parser.parse_args()
+
+  all_resources, application = expand_and_load_resources()
+
+  metadata = None
+  if args.metadata:
+    with open(args.metadata, "r") as metadatafile:
+      metadata = json.loads(metadatafile.read())
+  
+  if metadata:  
+    validate_metadata(metadata, application)
+
+  schema = load_yaml("/data/schema.yaml")
+  validate_images(schema, all_resources)
+
+
+def validate_images(schema, resources):
+  declared_images = set([deep_get(param, "default") for key, param in schema["properties"].iteritems() 
+    if deep_get(param, GOOGLE_MARKETPLACE, "type") == "IMAGE"])
+
+  used_images = []
+  for resource in resources:
+    used_images.extend(deep_find(resource, "image"))
+  used_images = set(used_images)
+
+  undeclared_images = [image for image in used_images 
+    if image not in declared_images]
+
+  if undeclared_images:
+    error_message = "ERROR Images should be declared in schema file."
+    print(error_message)
+    for image in undeclared_images:
+      print("  " + str(image))
+
+    raise ValidationException(error_message)
+
+  print("All {} used images are declared.".format(len(used_images)))
+
+  unused_images = [image for image in declared_images 
+    if image not in used_images]
+
+  if unused_images:
```

Comment:
this validation is not always desired. A partner has a use case in which no workloads use a declared image, but they would still want to declare it to be republished, and the value is actually contained in a CRD. Their application will create new workloads using the image from the CRD. Thus, the validation here is probably the set of extracted images must be a subset of the declared images.

We also want to do the following:
- Re-expand the manifests with new values for the images.
- Ensure that the extracted images also change (subset of the new values)
- This catches the case where the an image is declared and also hard coded to the default value.

This logic is non-trivial. I would suggest the following:
- Modify the unit test first to create this scenario. The current code should fail.
- Add a block of comment explaining what we're doing here.

Reply:
I don't follow after "We also want to do the following...". Lets sync offline.